cpp/src/join/unique_hash_join.cu

```diff
@@ -0,0 +1,386 @@
+
+    auto constexpr flushing_tile_size = cudf::detail::warp_size / window_size;
+    // random choice to tune
+    auto constexpr flushing_buffer_size = 2 * flushing_tile_size;
+    auto constexpr num_flushing_tiles   = UNIQUE_JOIN_BLOCK_SIZE / flushing_tile_size;
+    auto constexpr max_matches          = flushing_tile_size / tile_size;
+
+    auto const flushing_tile    = cg::tiled_partition<flushing_tile_size>(block);
+    auto const flushing_tile_id = block.thread_rank() / flushing_tile_size;
+
+    __shared__ cuco::pair<cudf::size_type, cudf::size_type>
+      flushing_tile_buffer[num_flushing_tiles][flushing_tile_size];
+    // per flushing-tile counter to track number of filled elements
+    __shared__ cudf::size_type flushing_counter[num_flushing_tiles];
+
+    if (flushing_tile.thread_rank() == 0) { flushing_counter[flushing_tile_id] = 0; }
+    flushing_tile.sync();  // sync still needed since cg.any doesn't imply a memory barrier
+
+    while (flushing_tile.any(idx < n)) {
+      bool active_flag = idx < n;
+      auto const active_flushing_tile =
+        cg::binary_partition<flushing_tile_size>(flushing_tile, active_flag);
+      if (active_flag) {
+        auto const found = hash_table.find(tile, *(iter + idx));
+        if (tile.thread_rank() == 0 and found != hash_table.end()) {
+          auto const offset = atomicAdd_block(&flushing_counter[flushing_tile_id], 1);
+          flushing_tile_buffer[flushing_tile_id][offset] = cuco::pair{
+            static_cast<cudf::size_type>(found->second), static_cast<cudf::size_type>(idx)};
+        }
+      }
+      flushing_tile.sync();
+      if (flushing_counter[flushing_tile_id] + max_matches > flushing_buffer_size) {
+        flush_buffer(flushing_tile,
+                     flushing_counter[flushing_tile_id],
+                     flushing_tile_buffer[flushing_tile_id],
+                     counter,
+                     build_indices,
+                     probe_indices);
+        flushing_tile.sync();
+        if (flushing_tile.thread_rank() == 0) { flushing_counter[flushing_tile_id] = 0; }
+        flushing_tile.sync();
+      }
+
+      idx += stride;
+    }  // while
+
+    flushing_tile.sync();
+    if (flushing_counter[flushing_tile_id] > 0) {
+      flush_buffer(flushing_tile,
+                   flushing_counter[flushing_tile_id],
+                   flushing_tile_buffer[flushing_tile_id],
+                   counter,
+                   build_indices,
+                   probe_indices);
+    }
+  }
+  // Scalar probing
+  else {
+    using block_scan = cub::BlockScan<cudf::size_type, UNIQUE_JOIN_BLOCK_SIZE>;
+    __shared__ typename block_scan::TempStorage block_scan_temp_storage;
+
+    auto constexpr buffer_capacity = 2 * UNIQUE_JOIN_BLOCK_SIZE;
+    __shared__ cuco::pair<cudf::size_type, cudf::size_type> buffer[buffer_capacity];
+    cudf::size_type buffer_size = 0;
+
+    while (idx - block.thread_rank() < n) {  // the whole thread block falls into the same iteration
```

Comment:
From above: `auto idx = cudf::detail::grid_1d::global_thread_id() / tile_size;`

Is this computation of `idx - block.thread_rank()` the correct way to index the thread-iterations, if `tile_size != 1`? It seems like you could get a different value than intended, and potentially deadlock at the `block.sync()` call...?

Reply:
This is the place I hesitated between `if constexpr` inside the kernel and two separate kernels. The `else` condition at line 201 corresponds to the `if constexpr` at line 142 thus this code path will be selected only when `tile_size == 1`. 

Sorry about the confusion with this long unreadable kernel. It is a testbench for the `cuco::static_set::retrieve` and eventually the kernel together with its device utilities will be removed once cuco `retrieve` is ready. We don't need to maintain this in libcudf.

Reply:
Thanks for the context. I had one small related comment here: https://github.com/rapidsai/cudf/pull/14990#discussion_r1488780061